File: client/mobile/types/conversation.ts
export interface CreateOnlineConversationRequest {
  novel?: string;
  shortStory?: string;
  poem?: string;
  play?: string;
  film?: string;
  writtenBy?: string;
  rule?: string;
  capacity: number;
  time: string;
  length: string;
}

export interface OnlineConversationDetailResponse {
  id: string;
  novel?: string;
  shortStory?: string;
  poem?: string;
  play?: string;
  film?: string;
  writtenBy?: string;
  rule?: string;
  capacity: number;
  time: string;
  length: string;
  canEnter: boolean;
  isRegistrant: boolean;
  moderatorIds: string[];
  isNotificationScheduled: boolean;
}

export interface BanParticipantRequest {
  conversationId: string;
  banId: string;
}

export interface ConversationSignalResponse {
  fromIds: string[];
  signal?: PeerSignal;
}

type PeerSignal =
  | { type: "offer" | "answer"; sdp: string }
  | { type: "candidate"; candidate: RTCIceCandidate }
  | { type: "leave" }
  | { type: "ban" }
  | { type: "mute" }
  | { type: "name-offer" | "name-answer"; name: string };

interface RTCIceCandidate {
  candidate?: string;
  sdpMLineIndex?: number | null;
  sdpMid?: string | null;
}

export type SeatCoordinate = { left: number; top: number };

export interface SeatAssignment {
  id?: string;
  name?: string;
  mute?: boolean;
  left: number;
  top: number;
}

export interface OfflineConversationMapResponse {
  id: string;
  writtenBy: string;
  lat: number;
  lng: number;
}

export interface OnlineConversationFeedResponse {
  id: string;
  novel: string;
  poem: string;
  shortStory: string;
  play: string;
  film: string;
  writtenBy: string;
  time: string;
}

export interface OfflineConversationSearchResponse {
  id: string;
  novel: string;
  poem: string;
  shortStory: string;
  play: string;
  film: string;
  writtenBy: string;
  time: string;
  lat: number;
  lng: number;
}

export interface CreateOfflineConversationRequest {
  novel?: string;
  shortStory?: string;
  poem?: string;
  play?: string;
  film?: string;
  writtenBy: string;
  rule?: string;
  time: string;
  length: number;
  mapsLink: string;
  location: string;
  city?: string;
  lat: number;
  lng: number;
  h3Res5: string;
  h3Res7: string;
}

export interface OfflineConversationDetailResponse {
  novel?: string | null;
  poem?: string | null;
  shortStory?: string | null;
  play?: string | null;
  film?: string | null;
  writtenBy: string;
  rule?: string | null;
  time: string;
  length: number;
  mapsLink: string;
  location: string;
  isModerator: boolean;
  isParticipant: boolean;
  numberOfParticipants: number;
  moderatorIds: string[];
}

export interface GetTurnResponse {
  uris: string[];
  username: string;
  credential: string;
}

File: backend/onlineconversation/internal/dto/conversation.go
package dto

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type CreateConversationRequest struct {
	Novel      string    `json:"novel"`
	ShortStory string    `json:"short_story"`
	Poem       string    `json:"poem"`
	Play       string    `json:"play"`
	Film       string    `json:"film"`
	WrittenBy  string    `json:"writtenBy"`
	Rule       string    `json:"rule"`
	Capacity   int       `json:"capacity"`
	Time       time.Time `json:"time"`
	Length     string    `json:"length"`
}

type OnlineConversationDetailResponse struct {
	Id                      uuid.UUID   `json:"id"`
	Novel                   string      `json:"novel,omitempty"`
	ShortStory              string      `json:"shortStory,omitempty"`
	Poem                    string      `json:"poem,omitempty"`
	Play                    string      `json:"play,omitempty"`
	Film                    string      `json:"film,omitempty"`
	WrittenBy               string      `json:"writtenBy"`
	Rule                    string      `json:"rule,omitempty"`
	Capacity                int         `json:"capacity"`
	Time                    time.Time   `json:"time"`
	Length                  string      `json:"length"`
	CanEnter                bool        `json:"canEnter"`
	IsRegistrant            bool        `json:"isRegistrant"`
	ModeratorIds            []uuid.UUID `json:"moderatorIds"`
	IsNotificationScheduled bool        `json:"isNotificationScheduled"`
}

type OnlineConversationDocument struct {
	Id         uuid.UUID `json:"id"`
	Novel      string    `json:"novel"`
	ShortStory string    `json:"short_story"`
	Poem       string    `json:"poem"`
	Play       string    `json:"play"`
	Film       string    `json:"film"`
	WrittenBy  string    `json:"writtenBy"`
	Time       time.Time `json:"time"`
}

type ConversationSignalResponse struct {
	FromIds []uuid.UUID     `json:"fromIds"`
	Signal  json.RawMessage `json:"signal,omitempty"`
}

type ConversationSignalRequest struct {
	ToIds  []uuid.UUID     `json:"toIds"`
	Signal json.RawMessage `json:"signal"`
}

type OnlineConversationFeedResponse struct {
	Id         uuid.UUID `json:"id"`
	Novel      string    `json:"novel,omitempty"`
	ShortStory string    `json:"shortStory,omitempty"`
	Poem       string    `json:"poem,omitempty"`
	Play       string    `json:"play,omitempty"`
	Film       string    `json:"film,omitempty"`
	WrittenBy  string    `json:"writtenBy"`
	Time       time.Time `json:"time"`
}

type BanParticipantRequest struct {
	ConversationId uuid.UUID `json:"conversationId"`
	BanId          uuid.UUID `json:"banId"`
}

type GetTurnResponse struct {
	Uris       []string `json:"uris"`
	Username   string   `json:"username"`
	Credential string   `json:"credential"`
}
